dandi/bids_validator_deno/__init__.py

```diff
@@ -0,0 +1,448 @@
+        stderr: str,
+        outfile_content: Optional[str] = None,
+    ):
+        """
+        Parameters
+        ----------
+        cmd : list[str]
+            The command that was executed as the execution of BIDS validator
+        returncode : int
+            The return code of the execution
+        stdout : str
+            The standard output of the execution
+        stderr : str
+            The standard error of the execution
+        outfile_content : Optional[str]
+            The content of the output file produced by the execution
+            (if any). This is `None` if the output file was not produced.
+        """
+        # Pass a human-readable message up to the base Exception
+        super().__init__("Execution of the deno-compiled BIDS validator failed")
+        self.cmd = cmd
+        self.returncode = returncode
+        self.stdout = stdout
+        self.stderr = stderr
+        self.outfile_content = outfile_content
+
+    def __str__(self):
+        base_msg = super().__str__()  # the message passed in __init__
+        return (
+            f"{base_msg}\n"
+            f"Command: `{' '.join(self.cmd)}`\n"
+            f"Return code: {self.returncode}\n"
+            f"Stdout:\n{self.stdout}\n"
+            f"Stderr:\n{self.stderr}"
+            + (
+                f"\nOutfile content:\n{self.outfile_content}"
+                if self.outfile_content is not None
+                else ""
+            )
+        )
+
+
+def strip_sgr(text: str) -> str:
+    """
+    Strip ANSI SGR (Select Graphic Rendition) sequences from a string.
+
+    For example, this can be used to remove color encoding sequences from terminal
+    outputs.
+    """
+    return _ANSI_SGR_PATTERN.sub("", text)
+
+
+def _invoke_validator(args: list[str]) -> CompletedProcess:
+    """
+    Invoke the deno compiled BIDS validator
+
+    Parameters
+    ----------
+    args : list[str]
+        An ordered list of options and arguments to pass to the validator
+
+    Returns
+    -------
+    CompletedProcess
+        An object representing the result of invoking the deno-compiled BIDS validator
+
+    Raises
+    ------
+    RuntimeError
+        If the deno-compiled BIDS validator times out in `TIMEOUT` seconds
+
+    Notes
+    -----
+    - The text captured in `stdout` and `stderr` can be color encoded
+    """
+    try:
+        result = run(
+            args=[CMD, *args],
+            capture_output=True,
+            timeout=TIMEOUT,
+            text=True,
+        )
+    except TimeoutExpired as e:
+        raise RuntimeError(
+            f"The `{' '.join(e.cmd)}` command timed out after {e.timeout} " f"seconds"
+        ) from e
+
+    return result
+
+
+def bids_validate(
+    dir_: DirectoryPath,
+    config: Optional[dict] = None,
+    ignore_nifti_headers: bool = False,
+    recursive: bool = False,
+) -> list[ValidationResult]:
+    """
+    Validate a file directory as a BIDS dataset with the deno-compiled BIDS validator
```

Comment:
The PR description states

> Breaking change:
dandi.validate.validate_bids() is removed and is replaced by `dandi.bids_validator_deno.bids_validate()`

But I do not see this removal, only a pure addition of `dandi.bids_validator_deno.bids_validate` (which I agree with; if removal is indeed a goal it can at least be done in a soft deprecation cycle at the very least)

So perhaps just the PR description is in need of tidying for bookkeeping purposes? Not a huge deal either way

Reply:
Thanks for pointing that out. We initially started with the goal of replacing the use of `bidsschematools` completely, but it turned out that it was not possible since the BIDS validator does not provide the same kind of metadata information, so we ended up keeping `dandi.validate.validate_bids()`, which provides an interface to `bidsschematools`.

The top post is now adjusted.

Regarding use of metadata gathered by `bidsschematools`, I think there are some problems there. I am currently leaving the related logic untouched since it is not relevant to this PR, but we will need to come back to it later.